backend/core/src/main/java/org/sonarsource/sonarlint/core/commons/SmartCancelableLoadingCache.java

```diff
@@ -0,0 +1,186 @@
+ */
+public class SmartCancelableLoadingCache<K, V> implements AutoCloseable {
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+
+  private final ExecutorService executorService;
+  private final String threadName;
+  private final BiFunction<K, SonarLintCancelChecker, V> valueComputer;
+  private final ConcurrentHashMap<K, ValueAndComputeFutures> cache = new ConcurrentHashMap<>();
+
+  private class ValueAndComputeFutures {
+    private final K key;
+    private CompletableFuture<V> valueFuture = new CompletableFuture<>();
+    private CompletableFuture<V> computeFuture;
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    public ValueAndComputeFutures(K key) {
+      this.key = key;
+      scheduleComputation();
+    }
+
+    private void scheduleComputation() {
+      lock.writeLock().lock();
+      if (computeFuture != null) {
+        computeFuture.cancel(false);
+        computeFuture = null;
+      }
+      // Even if we have canceled the previous computation, the previous valueFuture can still be completed later, so we will create a new one
+      // that will be returned to new callers. We will also try to complete the old valueFuture, in case the cancellation was successful.
+      var oldValueFuture = valueFuture;
+      valueFuture = new CompletableFuture<>();
+
+      var previousValue = oldValueFuture.getNow(null);
+
+      CompletableFuture<SonarLintCancelChecker> start = new CompletableFuture<>();
+      CompletableFuture<V> result = start.thenApplyAsync(cancelChecker -> valueComputer.apply(key, cancelChecker), executorService);
+      start.complete(new SonarLintCancelChecker(result));
+      result.whenCompleteAsync((newValue, error) -> {
+        lock.writeLock().lock();
+        try {
+          computeFuture = null;
+          if (error instanceof CancellationException) {
+            return;
+          }
+          try {
+            if (listener != null) {
+              listener.afterCachedValueRefreshed(key, previousValue, newValue);
+            }
+          } finally {
+            if (error != null) {
+              oldValueFuture.completeExceptionally(error);
+              valueFuture.completeExceptionally(error);
+            } else {
+              oldValueFuture.complete(newValue);
+              valueFuture.complete(newValue);
+            }
+          }
+        } finally {
+          lock.writeLock().unlock();
+        }
+      }, executorService);
+      computeFuture = result;
+      lock.writeLock().unlock();
+    }
+
+
+    public void refresh() {
+      scheduleComputation();
+    }
+
+    public CompletableFuture<V> getValueFuture() {
+      lock.readLock().lock();
+      try {
+        return valueFuture;
+      } finally {
+        lock.readLock().unlock();
+      }
+    }
+
+    public void cancel() {
+      lock.writeLock().lock();
```

Comment:
This means there is no possible cancelation while a computation is in progress, right?

Reply:
No, the write lock is only taken to schedule a new computation, since we might have to swap the completable future. The computation itself runs out of the lock.